### ISSUE TITLE
query: add `limit` modifier

### ISSUE THREAD

**Issue by dandv** [VictoriaMetrics/VictoriaMetrics#2467]
**Is your feature request related to a problem? Please describe.**
My application needs to [read more raw samples than fit in memory](https://stackoverflow.com/questions/71849368/return-a-limited-number-of-prometheus-samples-between-start-and-end-timestamps). The classic solution to this is pagination:
1. read a specific number number of samples in the memory, starting from the start of the desired range
2. process them
3. repeat step 1, starting from right after the last read sample

InfluxDB provides the [`LIMIT` clause](https://docs.influxdata.com/influxdb/v1.8/query_language/explore-data/#the-limit-and-slimit-clauses), which makes this simple and straightforward. [Prometheus does not have such a feature.](https://github.com/prometheus/prometheus/issues/10107#issuecomment-1098530662)

**Describe the solution you'd like**
A `limit` modifier

**Describe alternatives you've considered**
Use InfluxDB instead of VictoriaMetrics.

**Comment by hagen1778** [comment]
Hi @dandv! Have you checked [limit_offset](https://docs.victoriametrics.com/MetricsQL.html#limit_offset) function?

**Comment by dandv** [comment]
Yes, but as the docs state, it limits the number of series, not the number of raw samples returned. Test:

    curl https://play.victoriametrics.com/select/accounting/1/6a716b0f-38bc-4856-90ce-448fd713e3fe/prometheus/api/v1/query -d 'query=limit_offset(2, 0, process_cpu_seconds_total[1m])' | jq

returns 2 series:

```json
{
  "status": "success",
  "isPartial": false,
  "data": {
    "resultType": "vector",
    "result": [
      {
        "metric": {
          "__name__": "process_cpu_seconds_total",
          "cluster_num": "1",
          "cluster_retention": "1m",
          "instance": "play-1m-1-vmselect-2.us-east1-b.c.victoriametrics-test.internal:8481",
          "job": "vmselect"
        },
        "value": [
          1649936359,
          "155.2"
        ]
      },
      {
        "metric": {
          "__name__": "process_cpu_seconds_total",
          "cluster_num": "1",
          "cluster_retention": "admin",
          "instance": "play-admin-1-vmadmin-1.us-east1-b.c.victoriametrics-test.internal:8490",
          "job": "vmadmin"
        },
        "value": [
          1649936359,
          "106.49"
        ]
      }
    ]
  }
}
```

(Note that each series has only one sample, ignoring the `[1m]` duration. Not sure why that is, because called without `limit_offset`, the query returns 6 samples per series, and `limit_offset(2, 0, process_cpu_seconds_total)[10m]` returns 3 samples per series.)

**Comment by dandv** [comment]
Since this isn't a question, can we change the label to `enhancement` please?

**Comment by dandv** [comment]
Any updates on this? @dmitryk-dk & @valyala - this is a really basic feature. If it's not supposed to exist, it would be nice to document some workarounds.

**Comment by dmitryk-dk** [comment]
Hi @dandv ! We discussed your question and decided not to add this function to the MetricsQL because you can use the different workaround to implement the approach that you want to get:
1. You can query data and specify the time ranges (like start and end params) to get the exact limit of your raw series
2. You can limit the stream into your application by providing some read limiter

**Comment by dandv** [comment]
> 1. You can query data and specify the time ranges (like start and end params) to get the exact limit of your raw series

The problem is you can't know how many samples are in between `start` and `end`. The idea is to limit to N samples, which would also bring the benefit of constant-size buffers.

> 2. You can limit the stream into your application by providing some read limiter

That can chop off a sample in the middle.

**Comment by hagen1778** [comment]
> The problem is you can't know how many samples are in between start and end. The idea is to limit to N samples, which would also bring the benefit of constant-size buffers.

@dandv it seems not trivial from the TSDB perspective due to its distributed nature. Receiving samples from many shards and limiting them consistently (so request with the same offset could return identical list of samples each time) sounds complex to implement and gives nothing back to the most of monitoring scenarios. 
Until it becomes critical or has no workarounds, we'd prefer not to make processing on VM side more complex than it is.

The workaround in your case would be to implement[ streaming parsing of the json](https://en.wikipedia.org/wiki/JSON_streaming). There are plenty of implementations for various languages, but the idea is to limit the number of received entries on the client and do processing in batches. VM server will kindly wait for client to process the response in full. Or if this is not an option - response can be saved to the filesystem and processed afterwards.
